### ISSUE TITLE
Make Image pull timeout configurable

### ISSUE THREAD

**Issue by silver83** [aws/amazon-ecs-agent#2560]
### Summary
We sometimes notice a very high pull time, normally 5 minutes suddenly becoming 1 hour.

### Description
We'd like some mechanism to mitigate, possibly setting a custom timeout to fail the task and retry.

If we fail the task, I assume it will probably get rescheduled to the same host - wondering if there's a way to also get some event that we timed out, to allow customer to perform an action (replace the possibly bad host or similar)

The pull timeout seems to be hardcoded to 2 hours
https://github.com/aws/amazon-ecs-agent/blob/a250409cf5eb4ad84a7b889023f1e4d2e274b7ab/agent/dockerclient/timeout.go#L21

**Comment by fenxiong** [comment]
> We'd like some mechanism to mitigate, possibly setting a custom timeout to fail the task and retry.

The agent has a config env `ECS_IMAGE_PULL_INACTIVITY_TIMEOUT`, which configures an inactivity timeout for image pull. When a pull did not make any progress in an interval as specified in this env, the pull will fail and the task will fail as a result. Currently the default value for this env is 1 min for linux and 3 min for windows. Setting this to a smaller value can help fail the pull earlier in case a pull is slow. I'm wondering could you try using that and see if it mitigates the problem in your case?

**Comment by silver83** [comment]
If the default is 1 minute but we still took 1 hour to complete the download, I suspect lowering it would be inconsistent (hit or miss) ? 

We're looking for a concrete cap on the overall pull time - that's the process correlated with our customer's experience in this case, and the one we currently alarm on.

**Comment by silver83** [comment]
Adding to this - 
I’m assuming that a timeout will fail the “pull” operation, but will most likely retry it on the same host. In extreme cases (bad host or networking issue) we might end up putting ourselves into a failure loop. 

Thinking out loud, a smarter mechanism to address this might require a higher level TASK_PLACEMENT_TIMEOUT, meaning if the task itself didn't start successfully for some period, we provision a new task, up to some maximum number of PROVISIONING tasks. 

Resulting process might look something like

1. Pull should timeout after $PULL_TIMEOUT minutes. on timeout, retry locally $PULL_ATTEMPTS times at most.
2. After those retries
    a. No longer timeout - let current PULL operation to continue as long as it likes
    b. In parallel, start another task
    c. First task to successfully start wins, and the other can be stopped. The associated host might then also be eligible to descale.

**Comment by ubhattacharjya** [comment]
Hi,

The code change for making image pull timeout configurable (https://github.com/aws/amazon-ecs-agent/commit/4e2c222c8d78dead48c63c7e01c3e2a77970ad3d) will go out with the release of agent version 1.44.2

Thanks,
Utsa

**Comment by ubhattacharjya** [comment]
Hi,
The agent version 1.44.2 has been released. 
Closing this issue.

Regards,
Utsa
